### PR TITLE
Make diagnostics clearer for `?` operator

### LINKED ISSUE #71309: "try expression alternatives have incompatible types" is confusing wording
Simple example:
```rust
fn foo(x: Result<i32, ()>) -> Result<(), ()> {
    let y: u32 = x?;
    Ok(())
}
```
Compilation result (`rust version 1.44.0-nightly (52fa23add 2020-04-18)`):
```
error[E0308]: try expression alternatives have incompatible types
 --> src/lib.rs:2:18
  |
2 |     let y: u32 = x?;
  |                  ^^ expected `u32`, found `i32`
  |
help: you can convert an `i32` to `u32` and panic if the converted value wouldn't fit
  |
2 |     let y: u32 = x?.try_into().unwrap();
  |                  ^^^^^^^^^^^^^^^^^^^^^^
```

Parts of this diagnostic are great, showing the mismatched types (`u32` and `i32`) and even suggesting a valid possible fix.  But the wording of the error itself is confusing.  What "try expression alternatives"?  I don't see any alternatives.

Taking a glance at the compiler implementation, it's referring to the arms of the `match` block that `?` desugars into.  But it's confusing to talk about those, because:
- They're not visible to the user.
- Why would there be a type mismatch between the arms in the first place, if there are only two arms and one of them always ends in a `return` rather than producing a value?  Did type inference infer the error branch as having type `u32` due to the context?

  If so, the message is technically accurate, but unhelpful: from the user's point of view, the 'real' issue is a mismatch between the expected type of `x?` (namely `i32`) and the context.  As far as I can imagine, this is the *only* possible cause of a type mismatch between the arms of a desugared `?`.  So it seems like it would be fine to just change the error message to not mention the arms.

As a sidenote, if I take the `-Zunpretty=hir` output and try to compile that:
```rust
#![feature(try_trait)]
use std::ops::Try;
use std::convert::From;
fn foo(x: Result<i32, ()>) -> Result<(), ()> {
    let y: u32 = match Try::into_result(x) {
        Ok(val) => val,
        Err(err) => return Try::from_error(From::from(err)),
    };
    Ok(())
}
```

...oddly enough, I don't get a "match arms have incompatible types" error, but a more helpful one instead:
```
error[E0308]: mismatched types
 --> q.rs:6:20
  |
6 |         Ok(val) => val,
  |                    ^^^ expected `u32`, found `i32`
  |
help: you can convert an `i32` to `u32` and panic if the converted value wouldn't fit
  |
6 |         Ok(val) => val.try_into().unwrap(),
  |                    ^^^^
```

### PR DESCRIPTION
**Pull request by JohnTitor**
Fixes #71309
r? @estebank

### CODE REVIEW

File: src/test/ui/issues/issue-51632-try-desugar-incompatible-types.stderr [hunk 1]
```diff
@@ -1,9 +1,11 @@
-error[E0308]: try expression alternatives have incompatible types
+error[E0308]: `?` operator has incompatible types
   --> $DIR/issue-51632-try-desugar-incompatible-types.rs:8:5
    |
 LL |     missing_discourses()?
    |     ^^^^^^^^^^^^^^^^^^^^^ expected enum `std::result::Result`, found `isize`
    |
+   = note: `?` operator cannot convert from `isize` to `std::result::Result<isize, ()>`
+   = help: `?` operator unwraps `Result` or `Option` and propagates its `Err` or `None`
```
Comment (JohnTitor) [inline src/test/ui/issues/issue-51632-try-desugar-incompatible-types.stderr:8]: It's verbose?

Comment (Lonami) [reply src/test/ui/issues/issue-51632-try-desugar-incompatible-types.stderr:8]: In my opinion this last note is not necessary. The clarification with concrete types of the first note should be enough, and people may be using the `try_trait` feature wit custom types where this won't apply.

Comment (estebank) [reply src/test/ui/issues/issue-51632-try-desugar-incompatible-types.stderr:8]: I agree, I feel the second note is probably unnecessary, no-one is likely to try to use `?` without knowing that `Result` and `Option` are supported.

File: src/test/ui/issues/issue-71309.stderr [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+   |
+LL |     let y: u32 = x?;
+   |                  ^^ expected `u32`, found `i32`
+   |
+   = note: `?` operator cannot convert from `i32` to `u32`
+   = help: `?` operator unwraps `Result` or `Option` and propagates its `Err` or `None`
+help: you can convert an `i32` to `u32` and panic if the converted value wouldn't fit
+   |
+LL |     let y: u32 = x?.try_into().unwrap();
+   |                  ^^^^^^^^^^^^^^^^^^^^^^
```
Comment (estebank) [inline src/test/ui/issues/issue-71309.stderr:12]: The suggestion seems to be using the incorrect `Span`. It should be replacing the `?` with the new code.

**Comment by Dylan-DPC** [comment]
Closing this due to inactivity.

**Pull request closed without merging**